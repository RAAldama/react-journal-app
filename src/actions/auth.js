import { types } from "../types/types";
import { firebase, googleAuthProvider } from '../firebase/firebase-config';
import { uiFinishLoading, uiStartLoading } from "./ui";

export const startLoginEmailPassword = (email, password) => {
    return ( dispatch ) => {
        dispatch( uiStartLoading() );

        firebase.auth().signInWithEmailAndPassword(email, password).then(
            ({ user }) => {
                dispatch( login(user.uid, user.displayName) );
                dispatch( uiFinishLoading() )
            }
        ).catch(
            e => {
                console.log(e);
                dispatch( uiFinishLoading() );
            }
        )
    }
}

export const startRegisterEmailPassword = (email, password, username) => {
    return ( dispatch ) => {
        firebase.auth().createUserWithEmailAndPassword(email, password).then(
            async( {user} ) => {
                await user.updateProfile({ displayName: username });

                dispatch( login(user.uid, user.displayName) );
            }
        ).catch(
            e => console.log(e)
        )
    }
}

export const startLoginGoogle = () => {
    return ( dispatch ) => {
        firebase.auth().signInWithPopup( googleAuthProvider ).then(
            ( {user} ) => {
                dispatch( login(user.uid, user.displayName) );
            }
        );
    }
}

export const login = (uid, displayName) => {
    return {
        type: types.login,
        payload: {
            uid,
            displayName
        }
    }
}