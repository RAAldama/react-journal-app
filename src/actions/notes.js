import Swal from "sweetalert2";
import { db } from "../firebase/firebase-config";
import { loadNotes } from "../helpers/loadNotes";
import { types } from "../types/types";

export const startNewNote = () => {
    return async(dispatch, getSate) => {

        const uid = getSate().auth.uid;
        const newNote = {
            title: '',
            body: '',
            date: new Date().getTime()
        }

        const doc = await db.collection(`${uid}/journal/notes`).add( newNote );
        dispatch( activeNote(doc.id, newNote) );

    }
}

export const activeNote = (id, note) => {
    return {
        type: types.notesActive,
        payload: {id, ...note}
    }
}

export const startLoginNotes = (uid) => {
    return async (dispatch) => {

        const notes = await loadNotes(uid);
        dispatch ( setNotes(notes) );
        
    }
}

export const setNotes = (notes) => {
    return {
        type: types.notesLoad,
        payload: notes
    }
}

export const startSaveNote = (note) => {
    return async(dispatch, getSate) => {

        const uid = getSate().auth.uid;

        if(!note.url){
            delete note.url;
        }

        const note2Firestore = {...note};

        delete note2Firestore.id;

        await db.doc(`${uid}/journal/notes/${note.id}`).update(note2Firestore);

        dispatch( refreshNote(note.id, note) );
        Swal.fire('Nota salvada', note.title, 'success');

    }
}

export const refreshNote = (id, note) => {
    return {
        type: types.notesUpdated,
        payload: { id, note }
    }
}