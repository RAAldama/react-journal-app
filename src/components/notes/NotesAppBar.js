import React from 'react'

export const NotesAppBar = () => {
    return (
        <div className="notes__appbar">
            <span>18 de enero de 2021</span>
            <div>
                <button className="btn">
                    Imagen
                </button>

                <button className="btn">
                    Guardar
                </button>
            </div>
        </div>
    )
}
