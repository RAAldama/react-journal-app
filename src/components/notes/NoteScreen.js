import React, { useEffect, useRef } from 'react'
import { useDispatch, useSelector } from 'react-redux'

import { activeNote } from '../../actions/notes';
import { useForm } from '../../hooks/useForm';
import { NotesAppBar } from './NotesAppBar'

export const NoteScreen = () => {

    //El doble punto es para renombrar la variable sacada del selector
    const { active:note } = useSelector(state => state.notes);
    const [formValues, handleInputChange, reset] = useForm(note);
    const {title, body} = formValues;
    const activeId = useRef( note.id );
    const dispatch = useDispatch();

    useEffect(() => {

        if(note.id !== activeId.current){
            reset(note);
            activeId.current = note.id;
        }

    }, [note, reset])

    useEffect(() => {
        
        dispatch( activeNote(formValues.id, {...formValues}) );

    }, [formValues, dispatch])

    return (
        <div className="notes__main">
            <NotesAppBar />

            <div className="notes__content">
                <input 
                    className="notes__title-input" 
                    type="text" 
                    placeholder="Its called Lothric..."
                    name="title" 
                    value={title}
                    onChange={handleInputChange}
                />

                <textarea 
                    className="notes__text-area" 
                    placeholder="Did u get good?" 
                    name="body" 
                    value={body}
                    onChange={handleInputChange}
                >
                </textarea>

                {
                    note.url && 
                    <div className="notes__image">
                        <img src="https://images7.alphacoders.com/109/1098539.png" alt="note-img" />
                    </div>
                }
            </div>
        </div>
    )
}
