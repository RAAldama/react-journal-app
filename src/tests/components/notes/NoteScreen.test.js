import configureStore from 'redux-mock-store'
import thunk from 'redux-thunk'
import { mount } from "enzyme"
import { Provider } from "react-redux";

import { NoteScreen } from '../../../components/notes/NoteScreen';
import { activeNote } from '../../../actions/notes';

jest.mock('../../../actions/notes', () => ({
    activeNote: jest.fn(),
}))

const middlewares = [thunk];
const mockStore = configureStore(middlewares);
const initState = {
    auth: {
        uid: '23u84u9ejfisd',
        name: 'Albus'
    },
    ui: {
        loading: false,
        msgError: null
    },
    notes: {
        notes: [],
        active: {
            id: 1234,
            title: 'Hola',
            body: 'Mundo',
            date: 0
        }
    }
}

let store = mockStore(initState);
store.dispatch = jest.fn();

const wrapper = mount( 
    <Provider store={ store }>
        <NoteScreen /> 
    </Provider>
);

describe('tests on <NoteScreen />', () => {
    
    test('should render correctly', () => {
        expect(wrapper).toMatchSnapshot();
    })
    
    test('should iniate activeNote', () => {
        wrapper.find('input[name="title"]').simulate('change', {
            target: {
                name: 'title',
                value: 'Olis again OwO'
            }
        })

        expect( activeNote ).toHaveBeenLastCalledWith( 1234, {body: 'Mundo', title: 'Olis again OwO', id: 1234, date: 0} );
    })
    
})
