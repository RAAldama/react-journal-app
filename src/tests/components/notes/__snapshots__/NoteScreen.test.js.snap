// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`tests on <NoteScreen /> should render correctly 1`] = `
<div
  className="notes__main"
>
  <div
    className="notes__appbar"
  >
    <span>
      18 de enero de 2021
    </span>
    <input
      id="fileSelector"
      name="file"
      onChange={[Function]}
      style={
        Object {
          "display": "none",
        }
      }
      type="file"
    />
    <div>
      <button
        className="btn"
        onClick={[Function]}
      >
        Imagen
      </button>
      <button
        className="btn"
        onClick={[Function]}
      >
        Guardar
      </button>
    </div>
  </div>
  <div
    className="notes__content"
  >
    <input
      className="notes__title-input"
      name="title"
      onChange={[Function]}
      placeholder="Its called Lothric..."
      type="text"
      value="Hola"
    />
    <textarea
      className="notes__text-area"
      name="body"
      onChange={[Function]}
      placeholder="Did u get good?"
      value="Mundo"
    />
  </div>
  <button
    className="btn btn-danger"
    onClick={[Function]}
  >
    Borrar Nota
  </button>
</div>
`;
